### PR TITLE
LDAP Authentication

### LINKED ISSUE #42: login-ldap

### PR DESCRIPTION
**Pull request by awtkns**
closes #42 

Adds the the option for users to be authenticated with LDAP instead of local accounts

### CODE REVIEW

**Comment by awtkns** [comment]
Also tested on the CRC network with the LDAP server

File: backend/api/auth/ldap.py [hunk 1]
```diff
@@ -12,23 +14,27 @@ def __init__(self):
...
+        if config.LDAP_AUTH_ENABLED:
             assert Connection(self.server, auto_bind=True)
 
-    def authenticate(self, username, password):
+    def _get_connection(self, username, password) -> Connection:
         if not password:
-            return None
+            return False
 
         cn = self.user_base + username
```
Comment (asim-shrestha) [inline backend/api/auth/ldap.py:24]: what is cn?

File: backend/api/auth/ldap.py [hunk 2]
```diff
@@ -12,23 +14,27 @@ def __init__(self):
...
-        connection = Connection(self.server, cn, password)
+        return Connection(self.server, cn, password)
+
+    def authenticate(self, username: str, password: str) -> bool:
+        return self._get_connection(username, password).bind()
```
Comment (asim-shrestha) [inline backend/api/auth/ldap.py:28]: will the .bind() still work if _get_connection returns false?

File: backend/api/auth/ldap.py (hunk 1 above)
Comment (awtkns) [reply backend/api/auth/ldap.py:24]: Its LDAP specific terminology https://stackoverflow.com/questions/18756688/what-are-cn-ou-dc-in-an-ldap-search

**Review by asim-shrestha — APPROVED** [review]
Looks good, are we able to write tests for the login route though?

File: backend/api/auth/ldap.py (hunk 1 above)
Comment (awtkns) [reply backend/api/auth/ldap.py:24]: It is ldap specific terminology https://en.wikipedia.org/wiki/LDAP_Data_Interchange_Format#LDIF_fields

File: backend/api/auth/ldap.py (hunk 2 above)
Comment (asim-shrestha) [reply backend/api/auth/ldap.py:28]: And same thing with the user factory

**Comment by awtkns** [comment]
> Looks good, are we able to write tests for the login route though?

We could but to do that we would need to set up a test ldap server in ci

**Comment by asim-shrestha** [comment]
> > Looks good, are we able to write tests for the login route though?
> 
> We could but to do that we would need to set up a test ldap server in ci

gotcha

**Pull request merged**